packages/arb-upgrades/index.ts

```diff
@@ -0,0 +1,361 @@
+const ensureCleanGitTree = () => {
+  const tree = childProcess.execSync('git diff-index HEAD --').toString().trim()
+  const found = tree.split('\n').find(diff => diff.includes('contracts'))
+  if (found) {
+    throw new Error(
+      `You have local changes to ${found}. Commit/stash/ get rid of them`
+    )
+  }
+}
+
+export const initUpgrades = (
+  hre: HardhatRuntimeEnvironment,
+  rootDir: string
+) => {
+  const getQueuedUpdates = async (): Promise<{
+    path: string
+    data: QueuedUpdates
+  }> => {
+    const network = await hre.ethers.provider.getNetwork()
+    const path = `${rootDir}/deployments/${network.chainId}_queued-updates.json`
+    try {
+      const jsonBuff = readFileSync(path)
+      return { path, data: JSON.parse(jsonBuff.toString()) as QueuedUpdates }
+    } catch (err) {
+      if (err.code === 'ENOENT') {
+        console.log('New network; creating queued updates file')
+        writeFileSync(path, JSON.stringify({}))
+        return { path, data: {} }
+      } else {
+        throw err
+      }
+    }
+  }
+
+  const getDeployments = async (): Promise<{
+    path: string
+    data: CurrentDeployments
+  }> => {
+    const network = await hre.ethers.provider.getNetwork()
+    const path = `${rootDir}/deployments/${network.chainId}_current_deployment.json`
+    try {
+      const jsonBuff = readFileSync(path)
+      return {
+        path,
+        data: JSON.parse(jsonBuff.toString()) as CurrentDeployments,
+      }
+    } catch (err) {
+      if (err.code === 'ENOENT') {
+        console.log(
+          'New network; need to set up _current_deployments.json file'
+        )
+        throw err
+      } else {
+        throw err
+      }
+    }
+  }
+
+  const getBuildInfoString = async (contractName: string) => {
+    const contracts = (await hre.artifacts.getAllFullyQualifiedNames()).filter(
+      curr => curr.includes(contractName)
+    )
+    if (contracts.length !== 1) throw new Error('Contract not found')
+    const info = await hre.artifacts.getBuildInfo(contracts[0])
+    return JSON.stringify(info)
+  }
+
+  const deployLogic = async (
+    contractNames: ContractNames[] | ContractNames
+  ) => {
+    ensureCleanGitTree()
+
+    if (!Array.isArray(contractNames)) {
+      contractNames = [contractNames]
+    }
+    console.log('Deploying logic contracts for ', contractNames.join(','))
+    const signers = await hre.ethers.getSigners()
+    if (!signers.length) {
+      throw new Error(
+        'No signer - make sure a key is properly set (check hardhat config)'
+      )
+    }
+
+    const signer = signers[0]
+    console.log('Using signer', signer.address)
+
+    const { path, data: queuedUpdatesData } = await getQueuedUpdates()
+
+    for (const contractName of contractNames) {
+      console.log('Deploying new logic for ', contractName)
+
+      const contractFactory = (
+        await hre.ethers.getContractFactory(contractName)
+      ).connect(signer)
+      // TODO: ensure we're on the right layer?
```

Comment:
we can try to `getCode` for an arb-os precompile to see if we are actually in the L2

ie arbsys at `0x0000000000000000000000000000000000000064` returns `0xfe` in L2 but `0x` in L1

Reply:
we can infer if the contract is intended for L1 or L2 based on the first 2 characters of `ContractNames` (except for StandardArbERC20 which is always L2)